[issue-#39] It is now possible to download group of contacts.

charismatic-claire opened the pull request:
I everyone. I implemented a way to solve issue #39 .
Please review and let me know if I need to change this in order to make it really work.

skjnldsv commented:
If you want some example on how to do that, I added some infos on this comment: https://github.com/nextcloud/contacts/issues/39#issuecomment-426308099

I will gladely help you if you want to create this plugin (check with @brjhaverkamp as well) 
Thanks a lot for your attention on this issue! Tell me if you have any questions or need help (I'm on vacation though, so my answer delay can be a few days ;) )

charismatic-claire commented:
Hi,
I was expecting that kind of review, that's why I asked. If I get it right you say that the way I made the "Download" symbol appear is alright. Also the assigned "downloadGroup(group)" is not a problem. But the way the actual result is constructed is not suitable.

We would need a "CardDAV" plugin handling the request, based on an URL as an argument, right? And this plugin would be a part of the "Calendar" project and the router of this project would trigger the execution of the plugin, resulting in the creation of the requested *.vcf file. Right?

That way, the result would be computed on the server side and not on the client side by the browser evaluating the JavaScript expressions. Do I get it? By the way: For the use case I have in mind, it is important that more than one address book can be used. Say five address books contain users from ten different groups and I want to export all members of a certain group, regardless of the address book they are stored in.

skjnldsv commented:
> Hi,
> I was expecting that kind of review, that's why I asked. If I get it right you say that the way I made the "Download" symbol appear is alright. Also the assigned "downloadGroup(group)" is not a problem. But the way the actual result is constructed is not suitable.

I did not looked as I'm on vacation (shame on me for checking github ;) ) But looking at the code it should look ok yes :)

> We would need a "CardDAV" plugin handling the request, based on an URL as an argument, right? And this plugin would be a part of the "Calendar" project and the router of this project would trigger the execution of the plugin, resulting in the creation of the requested *.vcf file. Right?

Yes, basically (Not really sure) it would be a request with params (urls?, array of addressbooks and vcards within? pick one or find another alternative :wink: ) that is sent to the server. Not really related to the 'calendar' project like you said, but I'll assume you meant carddav/caldav/webdav? If so yes! :)


> That way, the result would be computed on the server side and not on the client side by the browser evaluating the JavaScript expressions. Do I get it? By the way: For the use case I have in mind, it is important that more than one address book can be used. Say five address books contain users from ten different groups and I want to export all members of a certain group, regardless of the address book they are stored in.

Exactly that yes! The js computation will be too heavy. We need to delegate the work to the server for that (should be easy for it since this is basically a simple string concat) :)

skjnldsv commented:
If you need anything regarding the server part or the way our plugins work, please come on irc on #nextcloud-dev, and ask your question! It's nice seeing more people on board :hugs: 
You can also shoot me an email on skjnldsv@protonmail.com if you have any question! :)

charismatic-claire commented:
I wrote "Calendar" and meant "Contacts". I'm sorry for that. I just wanted to know if I need to touch another App for that. That's why I asked. Sounds like no, I don't.

Well I don't know when I will find the time to implement this. But at least I have a better picture of what needs to be done. Maybe that also helps @brjhaverkamp if he is also working on it...

charismatic-claire commented:
Alright. Now the *.vcf files to be exported get indeed created on the server side. Unfortunately I wasn't able to do it using a "CardDAV" plugin. Instead I added a new route, added a method to the PageController and created a GroupExportService to construct my result. This service is the worst part: I used simple database queries to fetch the data and combined it into a string...

I think I need to understand the "dav" project and especially the idea of the ImageExportPlugin to avoid this mess. Anyway. It works. If you got some advice for me, go ahead. Oh and please enjoy your vacation first. No need to hurry.

skjnldsv commented:
> Alright. Now the *.vcf files to be exported get indeed created on the server side. Unfortunately I wasn't able to do it using a "CardDAV" plugin. Instead I added a new route, added a method to the PageController and created a GroupExportService to construct my result. This service is the worst part: I used simple database queries to fetch the data and combined it into a string...

This is looking close to what we need to do :)
Unfortunately (sorry :see_no_evil:), we really need to do that on https://github.com/nextcloud/server/ as a webdav plugin because this feature go further that being able to download contacts groups. Soon, we ill drop the CATEGORIES support and start using vcard4 groups, thus your implementation will no longer cover this feature. Implementing this as a carddav plugin will also allow us to add a checkbox support with selection and allow us to download any kind of contacts selection (groups, manual selection, search results...)
We also need to keep this app as simple as we can, therefore, all kind of php computation should be send off to server. So that it become globally available and not only when the contacts app is enabled :)

> I think I need to understand the "dav" project and especially the idea of the ImageExportPlugin to avoid this mess. Anyway. It works. If you got some advice for me, go ahead. Oh and please enjoy your vacation first. No need to hurry.

No need to rush for this feature yes. If you're okay with that, I can open a pull request with you on /server and we can start working this feature a bit? So that we could implement it on the next releases? :)
Ideally this should be shipped with nextcloud 15! ;)

charismatic-claire commented:
Ok, so we got a little misunderstanding here. When I asked:

> We would need a "CardDAV" plugin handling the request, based on an URL as an argument, right? And this plugin would be a part of the "Contacts" project and the router of this project would trigger the execution of the plugin, resulting in the creation of the requested *.vcf file. Right?

I specifically meant: Do I need to touch other apps than this one to achieve what I want? And I thought the answer was No, but the answer seems to be Yes. So... I think we should open a forum thread or something on this and discuss the requirements first. Only than it will be possible to implement exactly what we need, won't it?

Thanks for your quick reply again. I am learning a lot about Nextcloud here. Never contributed to it before, so... It's fun.

skjnldsv commented:
@charismatic-claire I'm so sorry for the misunderstood :stuck_out_tongue_closed_eyes: 
Well, we can discuss it on #39 if you want :)

charismatic-claire commented:
Okay. Here are some architectural thoughts and ideas:

- From the "contacts" app, we are only calling a function, say a `downloadGroup(group)` function via the computed property `groupsMenu()` in the `Contacts.vue` component.
- This call basically is a HTTP GET request from an URL like this `http://<SERVER_ADDRESS>/remote.php/dav/addressbooks/users/<USER_NAME>/<GROUP_NAME>?export`
- The actual work happens on the server side. For this we need to make a pull request to the "server" project itself.
- What we need is a `GroupExportPlugin.php` with namespace `OCA\DAV\CardDAV` extending the "dav" application of the Nextcloud server.
- The `GroupExportPlugin` would extend the `ServerPlugin` from `Sabre\DAV\ServerPlugin`.
- We would use `dav/appinfo/v1/carddav.php` to register our new plugin.
- First major question would be: How to implement the magic we want?
- Second major question would be: How and where to register the routes properly?

Well, is that more or less what we want, or do I absolutely not get it?

charismatic-claire commented:
Did it.

charismatic-claire commented:
Alright. Let me know if this is doable. Promises still confuse me, but I think I got it (almost) right...

skjnldsv commented on src/views/Contacts.vue:
```diff
@@ -313,6 +322,57 @@ export default {
+				})
+				// promise #3: merge into one object
+				.then(response => {
+					return response.map(data => data.body).join('')	// question: still to much heavy lifting in the browser?
```
Should be enough I think. We don't really have much other choice :)

skjnldsv commented on src/views/Contacts.vue:
```diff
@@ -313,6 +322,57 @@ export default {
+			}))
+			// group vcard urls by addressbook
+			const groupedUrls = _.mapValues(
+				_.groupBy(urls, 'addressbook'),
```
And you'll end up with an object  like this: :tada: 
```js
{
    test-1: {
        addressbook: {dav: {...}},
        contacts: [
            'contacts/url/1',
            'contacts/url/2'
        ]
    }
}
```

skjnldsv commented on src/views/Contacts.vue:
```diff
@@ -313,6 +322,57 @@ export default {
+							const addressbook = addressbooks.find(ab => ab.displayname === pair[0])
+							return addressbook.addressbookMultigetExport(pair[1])
+						})
+					return Promise.all(responses)
```
Then you should be able to do a map on all the addressbooks and do the same Promise.all. :)
```js
Promise.all(Object.keys(list).map(key => {
    return list[key].addressbook.dav.addressbookMultigetExport(list[key].contacts)
}))
```

skjnldsv replied:
And after concatenate the data you got, and trigger the download :)

skjnldsv commented on src/views/Contacts.vue:
```diff
@@ -313,6 +322,57 @@ export default {
+		 * @returns {Function} trigger download of resulting single vcard file
+		 */
+		downloadGroup(group) {
+			const contacts = this.contacts
```
Okay, a little help here :)
We already get the addressbooks from the store, so you do not need to findAllAddressBooks again.

Actually, if we pass the group object as the parametter directly, you can create an object with every addressbook and the associated contacts within:

```js
const list = {}
group.contacts.map((key) => {
	const id = this.contacts[key].addressbook.id
	list = Object.assign({
		[id]: {
			addressbook: this.contacts[key].addressbook,
			contacts: []
        }
	}, list)
	list[id].contacts.push(this.contacts[key].url)
})
```

skjnldsv commented on src/views/Contacts.vue:
```diff
@@ -145,7 +147,14 @@ export default {
+							{
+								icon: 'icon-download',
+								text: 'Download',
+								action: this.downloadGroup(group)
```
```js
groupsMenu() {
	const self = this
	...
			text: 'Download',
			action: function() {
				self.downloadGroup(group)
			}
```

skjnldsv commented on src/views/Contacts.vue:
```diff
@@ -313,6 +322,57 @@ export default {
+					return response.map(data => data.body).join('')	// question: still to much heavy lifting in the browser?
+				})
+			// download vcard
+			return function() {
```
Maybe create a dedicated function to download the data as a file, for readability? :)

skjnldsv commented on src/views/Contacts.vue:
```diff
@@ -152,7 +151,7 @@ export default {
 								icon: 'icon-download',
 								text: 'Download',
-								action: this.downloadGroup(group)
+								action: this.downloadGroup(group)		// does not work if called via 'self'. why is that?
```
You need to redefine self to this before the return declaration (line 139) :wink: 
`const self = this`

skjnldsv replied:
Also, you can add comments in github (like I'm doing right now) instead of in the code, it's easier to discuss. :hugs:

charismatic-claire replied:
`const self = this` was not the problem, I did that. My problem was that I had a function returning a function so I could not use a function as a reference. But now I can. Solved.

georgehrke commented on src/views/Contacts.vue:
```diff
@@ -145,7 +148,14 @@ export default {
+							{
+								icon: 'icon-download',
+								text: 'Download',
+								action: () => self.downloadGroup(group)
```
Since you are using an arrow function, you should be able to use `this` right away instead of `self`.
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions#No_separate_this

Did you try that? :)

georgehrke commented on src/views/Contacts.vue:
```diff
@@ -315,6 +325,56 @@ export default {
+			// create vcard promise with the requested contacts
+			const vcardPromise = Promise.all(
+				Object.keys(groupedContacts).map(key => {
+					return groupedContacts[key].addressbook.dav.addressbookMultigetExport(groupedContacts[key].contacts)
```
If there is only one command in the function and that's the return value, you can just use the concise body syntax:
```
Object.keys(groupedContacts).map(key =>
	groupedContacts[key].addressbook.dav.addressbookMultigetExport(groupedContacts[key].contacts)
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/Arrow_functions#Function_body

georgehrke commented on src/views/Contacts.vue:
```diff
@@ -315,6 +325,56 @@ export default {
+					return groupedContacts[key].addressbook.dav.addressbookMultigetExport(groupedContacts[key].contacts)
+				}))
+				.then(response => {
+					return {
```
Please use the concise body syntax here as well. Please note that you will have to put the object into parentheses in this case.

georgehrke commented:
Works very well, will give approval when the code remarks above have been addressed! :)

skjnldsv commented:
@charismatic-claire Could you also add your copyright in the headers of the files you edited please? :)

charismatic-claire commented:
@skjnldsv Since I only changed `Conacts.vue` in the end, there isn't much to do, right? Do you mean I should add something similar to your `@copyright Copyright (c) 2018 John Molakvoæ ...` in the header of this file? And should I also add something similar to your `@author John Molakvoæ ...`?

skjnldsv commented:
Leave the copyright, since I created this file originally. But add a new `@author` line with whatever entity you are representing for this pull request :)

Like the team popcorn did here: https://github.com/nextcloud/contacts/blob/c51d3e661a2d858907f36161fdb1042c075a58cf/src/components/Settings/SettingsAddressbook.vue#L5 :wink:

skjnldsv commented:
@charismatic-claire By the way! We're loving new contributors like you! We're meeting for new releases of nextcloud with a lot of contributors and we help a lot for the travel expenses! The next planned one is the week of the 11/03 in Stuttgart. We'd love to have you there! :wink: 
Also, I don't know what you do in life, if this is a Hobby or a work, but we have a new program specifically for supporting people of underrepresented groups in open source: https://nextcloud.com/include – you can apply for travel support for an event, mentorship or a lot of different ways we can help! :+1: 

Feel free to reach us! :hugs: 
Cheers, John.

charismatic-claire commented:
Hi John,

sorry, it has been a long time since I checked this thread. Well, thank you! I'm kind of proud that you said that and invited me to your next meeting.

> Also, I don't know what you do in life, if this is a Hobby or a work, but we have a new program specifically for supporting people of underrepresented groups in open source: https://nextcloud.com/include – you can apply for travel support for an event, mentorship or a lot of different ways we can help! 

Yeah about that: Actually it is my job. I'm a programmer. I wanted to learn JavaScript and Vue.js and a friend of mine needed this feature so I decided to just do it. Actually, I really want to learn more JavaScript and more Vue.js in the future, so I might stay engaged with Nextcloud. 

Thanks again for your kind way of supporting me with my first real contribution to an open source project!

skjnldsv commented:
@charismatic-claire awesome to hear! :tada: 
We love people like you who believe in open source projects and contributing not only to learn new things but because it will benefit for others as well! You missed the previous contributor week, but hold tight! We have a new one incoming! :wink: 
24/06/2019 -> 28/06/2019

We also have the annual nextcloud conference https://nextcloud.com/conf/
Which you're of course welcome to be part of! :+1:
